File: expression-problem/ocaml/expression_problem_1.ml

(* Demonstrates of a possible solution to expression problem in OCaml. *)

(* Classes *)

module type Document = sig
    type t
    val load : t -> unit
    val save : t -> unit
end

module type Printable = sig
    type t
    val print : t -> unit
end


module type PrintableDocument = sig
    type t = string
    include Document  with type t := t
    include Printable with type t := t
end


(* Text Document *)

module TextDocument: PrintableDocument = struct
    type t = string
    let load t = print_endline ("Loading TextDocument(" ^ t ^ ")")
    let save t = print_endline ("Savin TextDocument(" ^ t ^ ")")
    let print t = print_endline ("Printing TextDocument(" ^ t ^ ")")
end


(* Drawing Document *)

type drawing_document = Drawing_document of string

module DrawingDocument: PrintableDocument = struct
    type t = string
    let load t = print_endline ("Loading DrawingDocument(" ^ t ^ ")")
    let save t = print_endline ("Savin DrawingDocument(" ^ t ^ ")")
    let print t = print_endline ("Printing DrawingDocument(" ^ t ^ ")")
end



(* Demonstration *)

let test (module D: PrintableDocument) t =
    D.load t;
    D.save t;
    D.print t


let () =
    print_endline "";
    test (module TextDocument) "text";
    print_endline "";
    test (module DrawingDocument) "drawing";
    print_endline "";

File: expression-problem/haskell/expression-problem-1.hs

-- Demonstrates of a possible solution to expression problem in Haskell.

module Main where

import Prelude hiding (print)

-- Classes

class Document a where
    load :: a -> IO ()
    save :: a -> IO ()

class Printable a where
    print :: a -> IO ()


-- Text Document

data TextDocument = TextDocument String

-- Document Interface
instance Document TextDocument where
    load (TextDocument a) = putStrLn ("Loading TextDocument(" ++ a ++ ")...")
    save (TextDocument a) = putStrLn ("Saving TextDocument(" ++ a ++ ")...")

-- Printable Interface
instance Printable TextDocument where
    print (TextDocument a) = putStrLn ("Printing TextDocument(" ++ a ++ ")")


-- Drawing Document

data DrawingDocument = DrawingDocument String

-- Document Interface
instance Document DrawingDocument where
    load (DrawingDocument a) = putStrLn ("Loading DrawingDocument(" ++ a ++ ")...")
    save (DrawingDocument a) = putStrLn ("Saving DrawingDocument(" ++ a ++ ")...")

-- Printable Interface
instance Printable DrawingDocument where
    print (DrawingDocument a) = putStrLn ("Printing DrawingDocument(" ++ a ++ ")")


-- Demonstration

test a = do
    load a
    save a
    print a

main = do
    putStrLn ""
    test (TextDocument "text")
    putStrLn ""
    test (DrawingDocument "text")
    putStrLn ""

